crates/dojo/core/src/meta/introspect.cairo

```diff
@@ -722,3 +731,868 @@ pub impl Introspect_span<T, +Introspect<T>> of Introspect<Span<T>> {
         Ty::Array([Introspect::<T>::ty()].span())
     }
 }
+
+pub impl Introspect_FixedArray<T, const N: usize, +Introspect<T>> of Introspect<[T; N]> {
+    fn size() -> Option<usize> {
+        match Introspect::<T>::size() {
+            Option::Some(size) => Option::Some(size * N),
+            Option::None => Option::None,
+        }
+    }
+    fn layout() -> Layout {
+        Layout::FixedArray([(Introspect::<T>::layout(), N)].span())
+    }
+    fn ty() -> Ty {
+        Ty::FixedArray([(Introspect::<T>::ty(), N)].span())
+    }
+}
+
+pub impl IntrospectTupleSize0 of Introspect<()> {
+    #[inline(always)]
+    fn size() -> Option<usize> {
+        Option::Some(0)
+    }
+    #[inline(always)]
+    fn layout() -> Layout {
+        Layout::Tuple([].span())
+    }
+    #[inline(always)]
+    fn ty() -> Ty {
+        Ty::Tuple([].span())
+    }
+}
+
+pub impl IntrospectTupleSize1<E0, impl I0: Introspect<E0>> of Introspect<(E0,)> {
+    #[inline(always)]
+    fn size() -> Option<usize> {
+        I0::size()
+    }
+    #[inline(always)]
+    fn layout() -> Layout {
+        Layout::Tuple([I0::layout()].span())
+    }
+
+    #[inline(always)]
+    fn ty() -> Ty {
+        Ty::Tuple([I0::ty()].span())
+    }
+}
+
+pub impl IntrospectTupleSize2<
+    E0, E1, impl I0: Introspect<E0>, impl I1: Introspect<E1>,
+> of Introspect<(E0, E1)> {
+    #[inline(always)]
+    fn size() -> Option<usize> {
+        let mut it = array![I0::size(), I1::size()].into_iter();
+        if it.any(|x| x.is_none()) {
+            return Option::None;
+        }
+
+        let mut it = array![I0::size(), I1::size()].into_iter();
+        Some(it.fold(0, |acc, x| acc + x.unwrap()))
+    }
+    #[inline(always)]
+    fn layout() -> Layout {
+        Layout::Tuple([I0::layout(), I1::layout()].span())
+    }
+
+    #[inline(always)]
+    fn ty() -> Ty {
+        Ty::Tuple([I0::ty(), I1::ty()].span())
+    }
+}
+
+pub impl IntrospectTupleSize3<
+    E0, E1, E2, impl I0: Introspect<E0>, impl I1: Introspect<E1>, impl I2: Introspect<E2>,
+> of Introspect<(E0, E1, E2)> {
+    #[inline(always)]
+    fn size() -> Option<usize> {
+        let mut it = array![I0::size(), I1::size(), I2::size()].into_iter();
+        if it.any(|x| x.is_none()) {
+            return Option::None;
+        }
+
+        let mut it = array![I0::size(), I1::size(), I2::size()].into_iter();
+        Some(it.fold(0, |acc, x| acc + x.unwrap()))
+    }
```

Comment:
Extracting the code to a function to complement the optimization mentioned above could also reduce the size of contracts when using multiple tuple sizes. 👍 

```rust
fn sum_sizes(sizes: Array<Option<usize>>) -> Option<usize> {
    let mut total_size = 0;
    for s in sizes.into_iter() {
        if s.is_none() {
            return Option::None;
        }

        total_size += s.unwrap();
    }

    Option::Some(total_size)
}

// ...

fn size() -> Option<usize> {
    sum_sizes(array![I0::size(), I1::size(), I2::size(), ...])
}
```

Reply:
Ah yes, really good idea 👍 Thanks 🙏